go-selinux/pwalk.go

```diff
@@ -0,0 +1,95 @@
+// +build selinux,linux
+
+package selinux
+
+/* The code below was taken from
+ * https://github.com/stretchr/powerwalk/blob/master/walker.go
+ * and modified to use github.com/karrick/godirwalk
+ */
+
+import (
+	"runtime"
+	"sync"
+
+	"github.com/karrick/godirwalk"
+	"github.com/pkg/errors"
+)
+
+// Walk walks the file tree rooted at root, calling options.Callback
+// for each file or directory in the tree, including root. All errors
+// that arise visiting files and directories are filtered by Callback.
+// The output is non-deterministic.
+//
+// For each file and directory encountered, Walk will trigger a new Go routine
+// allowing you to handle each item concurrently. A maximum of twice the
+// runtime.NumCPU() Callback goroutines will be called at any one time.
+func Walk(root string, options *godirwalk.Options) error {
+	return WalkLimit(root, options, runtime.NumCPU()*2)
```

Comment:
I still think the number of threads should be customizable.

Reply:
It is (if you choose to use `WalkLimit()` function)